model/project_vars.go

```diff
@@ -441,46 +466,123 @@ func (projectVars *ProjectVars) MergeWithRepoVars(repoVars *ProjectVars) {
+		return "", "", errors.Wrapf(err, "validating parameter name for project variable '%s'", varName)
+	}
+
+	return paramName, paramValue, nil
+}
+
+// validParamBasename is a regexp representing the valid characters for a
+// parameter's base name (i.e. excluding any slash-delimited paths). Valid
+// characters for a basename are alphanumerics, underscores, dashes, and
+// periods.
+var validParamBasename = regexp.MustCompile(`^[a-zA-Z0-9_.-]+$`)
+
+// validateVarNameCharset verifies that a project variable name is not empty and
+// contains only valid characters. It returns an error if it it's empty or
+// contains invalid characters that are not allowed in a parameter name.
+func validateVarNameCharset(varName string) error {
+	if len(varName) == 0 {
+		return errors.Errorf("project variable name cannot be empty")
+	}
+	if !validParamBasename.MatchString(varName) {
+		return errors.Errorf("project variable '%s' contains invalid characters - can only contain alphanumerics, underscores, periods, and dashes", varName)
 	}
+	return nil
+}
 
-	for _, varToParam := range varsToParams {
-		if varToParam.Name == varName {
+// validateParamNameUnique verifies if the proposed parameter name to be used is
+// unique within a project. It returns an error if the parameter name
+// conflicts with an already existing parameter name.
+func validateParamNameUnique(pm ParameterMappings, varName, paramName string) error {
+	proposedBasename := parameterstore.GetBasename(paramName)
+	for _, m := range pm {
+		basename := parameterstore.GetBasename(m.ParameterName)
+		if m.Name == varName {
 			continue
 		}
-		if varToParam.ParameterName == paramName {
+		if basename == proposedBasename {
 			// Protect against an edge case where a different project var
 			// already exists that has the exact same candidate parameter name.
 			// Project vars must map to unique parameter names.
-			return "", errors.Errorf("parameter name '%s' for project variable '%s' conflicts with project variable '%s'", paramName, varName, paramName)
+			return errors.Errorf("parameter basename '%s' for project variable '%s' conflicts with existing one for project variable '%s'", proposedBasename, varName, m.Name)
 		}
+	}
+
+	return nil
+}
 
+// createParamBasenameForVar generates a unique parameter basename from a
+// project variable name.
+func createParamBasenameForVar(varName string) (string, error) {
+	paramName := varName
+
+	if strings.HasPrefix(varName, "aws") || strings.HasPrefix(paramName, "ssm") {
+		// Parameters cannot start with "aws" or "ssm", adding a prefix
+		// (arbitrarily chosen as an underscore) fixes the issue.
+		paramName = fmt.Sprintf("_%s", paramName)
 	}
 
 	return paramName, nil
 }
+
+// gzipCompressedParamExtension is the extension added to the parameter name to
+// indicate that the parameter value had to be compressed to fit within the
+// parameter length limit.
+const gzipCompressedParamExtension = ".gz"
```

Comment:
Will it be an issue in the future if someone names their parameter *.gz?

Reply:
You're right, it would be invalid. I added a bit of validation + a unit test to disallow project vars ending in `.gz`. None of the existing project vars have `.gz`, so adding the new restriction should be fine.